src/back/wgsl/writer.rs

```diff
@@ -1353,9 +1353,9 @@ impl<W: Write> Writer<W> {
                             back::vector_size_str(rows)
                         )?;
                     }
-                    TypeInner::Vector { size, .. } => {
+                    TypeInner::Vector { size, width, .. } => {
                         let vector_size_str = back::vector_size_str(size);
-                        let scalar_kind_str = scalar_kind_str(kind);
+                        let scalar_kind_str = scalar_kind_str(kind, convert.unwrap_or(width));
```

Comment:
I know you didn't write the code for `Expression::As` but this seems wrong, we are using the type of the inner expression instead of the current expression, but maybe I'm reading this wrong.

Reply:
I think so far the assumption was that we always bitcast between types with same widths (i.e. scalar to scalar, vector to vector, matrix to matrix) but this does indeed not work with different widths.

Looks like we also do this in all the other backends. Will open an issue.